### PR TITLE
Sea Turtles - Joanna Dudley

### PR DESCRIPTION
**Pull request by jfdudley**

### CODE REVIEW

File: swap_meet/clothing.py [hunk 1]
```diff
@@ -1,2 +1,39 @@
...
+    age_description():
+        Returns string describing item's age
+    
+    ...
+    Methods:
+
+    __str__:
+        represents class objects as a string.
+    
+    """
```
Comment (kelsey-steven-ada) [inline swap_meet/clothing.py:32]: Great to see you getting practice with docstrings! While thorough, these are a little bit long for classes that won’t see extremely wide distribution. Verbose docstrings are more costly to keep up to date as code changes, so simpler docstrings are often easier to maintain over time. The usage of formal docstrings varies wildly in production teams, so you should be ready to adapt your practices to whatever norms the rest of the team has established.

File: swap_meet/clothing.py [hunk 2]
```diff
@@ -1,2 +1,39 @@
...
+    
+    def __init__(self, condition = 0, age = None):
+        super().__init__(condition = condition, age = age, category = "Clothing")
```
Comment (kelsey-steven-ada) [inline swap_meet/clothing.py:35]: Nice choice to make sure no one can pass a different category to the children classes. We always want a Clothing object to have the category "Clothing", so there's no reason to even give the option 😄

File: swap_meet/item.py [hunk 3]
```diff
@@ -1,2 +1,69 @@
...
+        elif self.condition <= 2:
+            return "It's....tolerable. I suppose. If one must."
+        elif self.condition <= 3:
+            return "Decent."
+        elif self.condition <= 4:
+            return "Alright, NOW we're talking. Nothing's perfect, right?"
+        elif self.condition <= 5:
+            return "Excellent. Full marks. Drinks all around!"
+        else:
+            return "Oh my god BUY THIS NOW ITS CONDITION IS OUT OF THIS WORLD"
```
Comment (kelsey-steven-ada) [inline swap_meet/item.py:54]: I believe there's an off by one error with this if-statement tree. Assuming an Item's condition will always be between 0-5, the description `"Oh my god BUY THIS NOW ITS CONDITION IS OUT OF THIS WORLD"` will never get returned. When an item has a `condition` of 5, it will hit the statement `elif self.condition <= 5:` and return `"Excellent. Full marks. Drinks all around!"` We would need to use a `<` operator on each of the if/elif statements instead of `<=` to see the results we expect.

Another option that could help prevent this error would be to make a tradeoff for memory and create a dictionary that maps integers/`condition` as keys to descriptions as values.  
```python
condition_map = {
    0: "I mean I wouldn't touch this with a ten foot pole but, uh, you do you.",
    1: "It's....tolerable. I suppose. If one must.",
    2: "Decent.",
    3: "Alright, NOW we're talking. Nothing's perfect, right?",
    4: "Excellent. Full marks. Drinks all around!",
    5: "Oh my god BUY THIS NOW ITS CONDITION IS OUT OF THIS WORLD"
}
        
return condition_map[int(self.condition)]
```

File: swap_meet/item.py [hunk 4]
```diff
@@ -1,2 +1,69 @@
...
+    
+    def age_description(self):
+        """Returns string describing item's age"""
+
+        if self.age is None:
+            return "The age of this item is unknown. Please judge it by its condition."
+        elif self.age == 0:
+            return "This item is brand new."
+        elif 0 < self.age < 1:
```
Comment (kelsey-steven-ada) [inline swap_meet/item.py]: This comparison could be simplified to `elif self.age < 1:` since the `elif` above handles the case if `self.age` is 0.

File: swap_meet/electronics.py [hunk 5]
```diff
@@ -1,2 +1,39 @@
...
+    
+    ...
+    Methods:
+
+    __str__:
+        represents class objects as a string.
+
+    """
+    
+    def __init__(self, condition = 0, age = None):
```
Comment (kelsey-steven-ada) [inline swap_meet/electronics.py:34]: The subclasses look great!

File: swap_meet/item.py [hunk 6]
```diff
@@ -1,2 +1,69 @@
...
+    age_description():
+        Returns string describing item's age
+
+    """
+    
+    def __init__(self, condition = 0, age = None, category = ""):
+        self.condition = condition
+        self.category = category
+        if age and age < 0:
+            raise ValueError("Age may not be a negative number")
```
Comment (kelsey-steven-ada) [inline swap_meet/item.py:33]: Nice check for invalid age data!

File: swap_meet/vendor.py [hunk 7]
```diff
@@ -1,2 +1,160 @@
...
+    swap_by_newest(other):
+        Swaps newest items of two instances' inventories and returns True.
+        - Returns False if either or both inventory items' ages are all unknown.
+    
+    """
+
+    def __init__(self, inventory = None):
+        if inventory is None:
+            inventory = []
+        self.inventory = inventory
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py]: Nice handling of the `None` default value! I generally stay away from overwriting the data we get from our parameters in case we need that data back. In this case there's no harm, but if we wanted to avoid overwriting the `inventory` variable we could do something like:
```python
if inventory is None:
    self.inventory = []
else:
    self.inventory = inventory
```
Or we could write this is with a ternary operator:
```python
self.inventory = inventory if inventory else []
```
This can be read as "If the parameter inventory is truthy, assign self.inventory to inventory, else assign self.inventory to an empty list". The general blueprint for a ternary operator looks like: `variable_name = <value_if_condition_true> if <condition_statement> else <value_if_condition_false>`.

The solution above will ignore if the user passes in an empty list (since an empty list is falsy) and create a new empty list. If we're worried about space and want to avoid that, we could be more explicit with our check for `None`
```python
self.inventory = inventory if inventory is not None else []
```
More info: https://book.pythontips.com/en/latest/ternary_operators.html

File: swap_meet/vendor.py [hunk 8]
```diff
@@ -1,2 +1,160 @@
...
+        return new_item
+    
+    def remove(self, discarded_item):
+        """Remove an item from inventory"""
+
+        if discarded_item not in self.inventory:
+            return False
+        else: 
+            self.inventory.remove(discarded_item)
+            return discarded_item
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py]: I like the choice to check if the item can be removed first. Since that if-check is at the top, we could remove the else and outdent the main logic (the intended operations if all data is valid) of the function:
```suggestion
        if discarded_item not in self.inventory:
            return False

        self.inventory.remove(discarded_item)
        return discarded_item
```

File: swap_meet/vendor.py [hunk 9]
```diff
@@ -1,2 +1,160 @@
...
+    
+    def get_by_category(self, category_string):
+        """Return list of items from inventory of a certain category"""
+
+        item_list = []
+        for item in self.inventory:
+            if item.category == category_string:
+                item_list.append(item)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py]: Notice this pattern of: 
```python
result_list = []
for element in source_list:
    if some_condition(element):
        result_list.append(element)
```

can be rewritten using a list comprehension as:
```python
result_list = [element for element in source_list if some_condition(element)]
```

Which here would look like:
```python
item_list = [item for item in self.inventory if item.category == category]
```

At first, this may seem more difficult to read, but comprehensions are a very common python style, so try getting used to working with them!

File: swap_meet/vendor.py [hunk 10]
```diff
@@ -1,2 +1,160 @@
...
+        if my_item not in self.inventory or their_item not in other.inventory:
+            return False
+        else:
+            # remove items from original inventory
+            self.remove(my_item)
+            other.remove(their_item)
+            # add items to new inventory 
+            self.add(their_item)
+            other.add(my_item)
+            return True
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py]: Similar to the feedback in the `remove` function, since the else-block of an if-statement is optional, we could remove the else and outdent the main logic of the function:
```suggestion
        if my_item not in self.inventory or their_item not in other.inventory:
            return False
            
        # remove items from original inventory
        self.remove(my_item)
        other.remove(their_item)
        # add items to new inventory 
        self.add(their_item)
        other.add(my_item)
        
        return True
```

File: swap_meet/vendor.py [hunk 11]
```diff
@@ -1,2 +1,160 @@
...
+        for item in self.inventory:
+            if item.category == category_string:
+                item_list.append(item)
+        return item_list
+    
+    def swap_items(self, other, my_item, their_item):
+        """Swaps items betwen self and another vendor's inventory"""
+
+        # check valid input
+        if my_item not in self.inventory or their_item not in other.inventory:
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:93]: Nice check for invalid data in one if-statement 🙂

File: swap_meet/vendor.py [hunk 12]
```diff
@@ -1,2 +1,160 @@
...
+    def swap_first_item(self, other):
+        """Swaps first item in self and another vendor's inventory"""
+
+        # check valid input
+        if not self.inventory or not other.inventory:
+            return False
+        else:
+            my_first_item = self.inventory[0]
+            vendor_first_item = other.inventory[0]
+            self.swap_items(other, my_first_item, vendor_first_item)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py]: Nice code reuse!

Something to consider: by using `swap_items`, this winds up being O(n) (mostly due to removing the items from the inventory). If it was important to lower the Big O complexity, we could swap the items in place here to have O(1) runtime with something like:
```python
self.inventory[0], other.inventory[0] = other.inventory[0], self.inventory[0]
```

File: swap_meet/vendor.py [hunk 13]
```diff
@@ -1,2 +1,160 @@
...
+            return True
+    
+    def get_best_by_category(self, desired_category):
+        """Returns item in best condition within a certain category"""
+
+        # get list of items of category
+        potential_items = self.get_by_category(desired_category)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:128]: Nice code reuse!

File: swap_meet/vendor.py [hunk 14]
```diff
@@ -1,2 +1,160 @@
...
+        # account for no items of desired category or only one of desired category
+        if not potential_items:
+            return None
+        elif len(potential_items) == 1:
+            return potential_items[0]
+        # calculate max item condition
+        else:
+            best_item = max(potential_items, key = lambda item: item.condition)
+            return best_item
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py]: I'm not sure we need to separate out the case where the list only contains one item. What do we gain by handling the logic for a single item in the list separately from lists of 2 or more items?

If we can combine cases, we can simplify the code a little:
```suggestion
        if not potential_items:
            return None
            
        best_item = max(potential_items, key = lambda item: item.condition)
        return best_item
```

File: swap_meet/vendor.py [hunk 15]
```diff
@@ -1,2 +1,160 @@
...
+            other.add(my_item)
+            return True
+    
+    def swap_first_item(self, other):
+        """Swaps first item in self and another vendor's inventory"""
+
+        # check valid input
+        if not self.inventory or not other.inventory:
+            return False
+        else:
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py]: I'm noticing that an else containing the main logic of the function is a pattern across the file so I'm not going to call out each instance. Something to think about going forward when writing a function, if an else is not required, try removing it to simplify the code. If the the main logic of a function is in an else, consider why and if that else can be removed so that the main logic is as visible and up front as possible.

File: swap_meet/vendor.py [hunk 16]
```diff
@@ -1,2 +1,160 @@
...
+        if not potential_items:
+            return None
+        elif len(potential_items) == 1:
+            return potential_items[0]
+        # calculate max item condition
+        else:
+            best_item = max(potential_items, key = lambda item: item.condition)
+            return best_item
+
+    def swap_best_by_category(self, other, my_priority, their_priority):
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:140]: Great approach to this function!

File: swap_meet/vendor.py [hunk 17]
```diff
@@ -1,2 +1,160 @@
...
+        for item in self.inventory:
+            if item.age:
+                potential_items.append(item)
+        if not potential_items:
+            return None
+        elif len(potential_items) == 1:
+            return potential_items[0]
+        else:
+            newest_item = min(potential_items, key = lambda item: item.age)
+            return newest_item
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py]: We could use the `filter` function or a list comprehension here to simplify our code that finds items with an age set:
```suggestion
        """Returns newest item from inventory items with known ages"""
        # Get all items with a valid age
        potential_items = [item for item in self.inventory if item.age]
        
        if not potential_items:
            return None
        
        # Find and return an item with the smallest age value
        return min(potential_items, key = lambda item: item.age)
```

File: swap_meet/vendor.py [hunk 18]
```diff
@@ -1,2 +1,160 @@
...
+                potential_items.append(item)
+        if not potential_items:
+            return None
+        elif len(potential_items) == 1:
+            return potential_items[0]
+        else:
+            newest_item = min(potential_items, key = lambda item: item.age)
+            return newest_item 
+    
+    def swap_by_newest(self, other):
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:174]: Nice improvements & use of helper functions for them!

File: tests/unit_tests/test_wave_01.py [hunk 19]
```diff
@@ -49,7 +49,4 @@ def test_removing_not_found_is_false():
 
     result = vendor.remove(item)
 
-    raise Exception("Complete this test according to comments below.")
-    # *********************************************************************
-    # ****** Complete Assert Portion of this test **********
-    # *********************************************************************
+    assert result is False
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_01.py:52]: This assert is great for confirming the return value is correct, but what else could we assert to make sure there were no side affects like changes to the inventory?

File: swap_meet/vendor.py [hunk 20]
```diff
@@ -1,2 +1,160 @@
...
+        # get list of items of category
+        potential_items = self.get_by_category(desired_category)
+        # account for no items of desired category or only one of desired category
+        if not potential_items:
+            return None
+        elif len(potential_items) == 1:
+            return potential_items[0]
+        # calculate max item condition
+        else:
+            best_item = max(potential_items, key = lambda item: item.condition)
```
Comment (kelsey-steven-ada) [inline swap_meet/vendor.py:137]: Great use of `max` with a lambda to simplify finding the item with the highest condition!

File: tests/unit_tests/test_wave_02.py [hunk 21]
```diff
@@ -34,7 +34,7 @@ def test_get_no_matching_items_by_category():
 
     items = vendor.get_by_category("electronics")
 
-    raise Exception("Complete this test according to comments below.")
-    # *********************************************************************
-    # ****** Complete Assert Portion of this test **********
-    # *********************************************************************
+    assert len(items) == 0
+    assert item_a and item_b and item_c not in items
+    assert items == []
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_02.py]: I would say both the asserts `assert len(items) == 0` and `assert items == []` check the same thing (whether `items` is empty). Since we want to know that it is a list and it is empty, we could just keep the second check `assert items == []`. 

Aside from confirming the `items` list is empty, what could we assert to make sure the `vendor`'s inventory was not changed?

File: tests/unit_tests/test_wave_06.py [hunk 22]
```diff
@@ -75,17 +75,16 @@ def test_swap_best_by_category():
...
-    # - That tai and jesse's inventories are the correct length
-    # - That all the correct items are in tai and jesse's inventories, including the items which were swapped from one vendor to the other
-
-@pytest.mark.skip
+    assert result == True
+    assert len(tai.inventory) and len(jesse.inventory)== 3
+    assert item_c in jesse.inventory and item_c not in tai.inventory
+    assert item_f in tai.inventory and item_f not in jesse.inventory
+    assert item_d and item_e in jesse.inventory
+    assert item_a and item_b in tai.inventory
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_06.py:85]: Great checks for all the relevant data!

File: tests/unit_tests/test_wave_07.py [hunk 23]
```diff
@@ -0,0 +1,267 @@
...
+
+    # Act
+    result = tai.swap_by_newest(jesse)
+
+    # Assert
+
+    assert not result
+    assert len(tai.inventory) and len(jesse.inventory) == 3
+    assert item_a and item_b and item_c in tai.inventory
+    assert item_d and item_e and item_f in jesse.inventory
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_07.py]: Really solid test suite for the age additions, great exploration of edge cases to check.

File: tests/unit_tests/test_wave_07.py [hunk 24]
```diff
@@ -0,0 +1,267 @@
...
+from swap_meet.decor import Decor
+from swap_meet.electronics import Electronics
+
+def test_age_condition_same_across_classes():
+    items = [
+        Clothing(age=5),
+        Decor(age=5),
+        Electronics(age=5)
+    ]
+    five_age_description = items[0].age_description()
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_07.py]: Something to consider: do we want to test that all items with the same age return the same description, or do we want to check if the description returned for a particular age is the string we expect it to be? 

For the condition description, the tests that came with the project were only checking the former since students wrote their own descriptions. In a well-used product, we would probably want to check both scenarios. What could that look like?

File: tests/unit_tests/test_wave_07.py [hunk 25]
```diff
@@ -0,0 +1,267 @@
...
+
+    items[0].age = 1
+    one_age_description = items[0].age_description()
+    assert isinstance(one_age_description, str)
+
+    for item in items:
+        item.age = 1
+        assert item.age_description() == one_age_description
+
+    assert one_age_description != five_age_description
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_07.py]: In this test, the arrange/act/assert step are interleaved and it is testing a few different things in one. This test checks:
- If the age description is a string
- If the description is the same for all items with the same age
- If the descriptions for age 1 and age 5 are different (Side note, what about the other age descriptions – would we want to confirm they are distinct as well?)

I would recommend breaking this test apart into 2-3 smaller tests with clear arrange/act/assert steps and a single focus for the test.

**Review by kelsey-steven-ada — COMMENTED** [review]
🎉 You completed Unit 1! 🎉 Great work on this project, I've left some suggestions for refactoring. Take a look, and let me know if anything needs clarification!

File: swap_meet/item.py (hunk 3 above)
Comment (jfdudley) [reply swap_meet/item.py:54]: Ah, I think I had missed or forgot that condition would necessarily always be between 0 and 5, and so was trying to account for the edge case that someone put in a higher or lower number, even if typical usage was to have condition in that range. 

In the condition map situation, if someone puts a float in, does the int() round the condition up or down (say, 4.2 resolves to 4 but 4.6 resolves to 5), or will it just resolve to the integer portion of the float (both 4.2 and 4.6 resolving to 4)?

File: swap_meet/vendor.py (hunk 10 above)
Comment (jfdudley) [reply swap_meet/vendor.py]: I think I'm just such a completionist that leaving the else bit off of an if else statement feels "wrong" still 😅. I'll work on it!

File: swap_meet/vendor.py (hunk 12 above)
Comment (jfdudley) [reply swap_meet/vendor.py]: This whole tuple swap thing that came up here in class as well is something I'd like to talk about at our next meeting! I'm just so perplexed as to what is happening under the hood that python knows what to do with just this expression and it's still confusing to me. I'd like to understand it better!

File: swap_meet/vendor.py (hunk 14 above)
Comment (jfdudley) [reply swap_meet/vendor.py]: I think I was figuring if there was just the one item it would save time (and maybe space as well) to not run a more complicated calculation to figure out the max - with only one option, it is by default the max. But perhaps that's not the case, depending on how the calculation works? Or maybe it just doesn't save enough time to make it worth it.

File: swap_meet/vendor.py (hunk 20 above)
Comment (jfdudley) [reply swap_meet/vendor.py:137]: Using max() with a key was fresh in my mind from our discussions on viewing party, so I was happy to get a chance to put it into practice!!

File: swap_meet/item.py (hunk 3 above)
Comment (kelsey-steven-ada) [reply swap_meet/item.py:54]: `int` will do the latter, it will chop off the decimals and just give you the integer. So the solution above works if you want everything from 4.0-4.99 to have the same text. If you wanted to return the text for whichever whole number the decimal is closest to, the `round()` function would do that.

File: swap_meet/vendor.py (hunk 12 above)
Comment (kelsey-steven-ada) [reply swap_meet/vendor.py]: Definitely, let's talk it through! High-level, what's happening is python is assigning multiple variables at the same time (there are many languages that allow this). We could use the same syntax if we need to assign multiple strings or any other type at the same time:
```python
# After this line, var_a will point to "value 1" and var_b will point to "value 2"
var_a, var_b = "value 1", "value 2" 

# After this line, int_a will point to 5, int_b will point to 12, and int_c will point to 20
int_a, int_b, int_c = 5, 12, 20
```
As you can see in the int example above, we're also not limited to only assigning 2 variables at a given time. We often choose not to do this multiple assignment because it can make code less readable by making it harder to see where an individual variable is declared, but for a direct swap operation like we're solving in Swap-Meet, it becomes really handy!

If we look back at the Swap-Meet code:
- On the left side we declare the variables we are assigning, `self.inventory[0]` and  `other.inventory[0]`. In this case instead of named variables they are locations in our vendor inventory lists, which like named variables are pointers to data. 

- On the right side we say what values the variables are being assigned, in the same order as the variable names are listed. In this case, instead of a raw piece of data or a named variable, the values being assigned are the Items pointed to by the first indices in our vendor inventories `other.inventory[0]` and `self.inventory[0]`.

As I understand it, python will look up the values to be assigned on the right of the `=` operator, then perform the assignment of those values to the variables on the left of the `=`.

File: swap_meet/vendor.py (hunk 14 above)
Comment (kelsey-steven-ada) [reply swap_meet/vendor.py]: That's a great consideration to have, I think it'd be worth writing a short comment to that effect above the `elif` to bring a little clarity. In this project we don't need the fast access if there's a single item, but there are many applications where it's important to be as performant as possible, and we may absolutely want to do something like this.

File: swap_meet/vendor.py (hunk 12 above)
Comment (jfdudley) [reply swap_meet/vendor.py]: > As I understand it, python will look up the values to be assigned on the right of the = operator, then perform the assignment of those values to the variables on the left of the =.

This makes sense thank you!!!! That was, I think, the phrasing I needed.

And this is what happens instantaneously, I'm guessing, so we don't get a similar issue as if we were changing a list we were iterating on? It's not like we change self.inventory[0] and then when we look there to change other.inventory[0] it's the new value - It checks the original values of both locations, and then assigns the new locations to those values?

Comment (kelsey-steven-ada) [reply swap_meet/vendor.py]: Exactly! If python instead tried to resolve the multiple assignments one after the other, in Swap-Meet we'd see a bug where the first list would get changed as we expect, but the second list would appear to not have been updated.

Comment (kelsey-steven-ada) [reply swap_meet/vendor.py]: If you're interested, this blog has some really nice examples of how wild multiple assignment can get and different ways it can be used: https://note.nkmk.me/en/python-multi-variables-values/

Comment (jfdudley) [reply swap_meet/vendor.py]: Thank you!!! This was super helpful. Sometimes you just need something phrased in a certain way for it to click!!
